File: bankomat/bankomat.ts
const fs = require('fs');

function readData() {
    return require('./accounts.json');
}

function writeData(a:any) {fs.writeFile('./accounts.json', JSON.stringify(a), (err:any) => {
    if (err){
        console.log(err)
    }
})}

interface OperationData {
    id: number;
    date: any;
    amount: number;
  }

interface AssignedCardsData{
    id: number;
    pin: number;
  }

interface UserData {
    id: number;
    first_name: string;
    last_name: string;
    email: string;
    gender: string;
    yearOfBorn: number;
    operations: OperationData[];
    assignedCards: AssignedCardsData[];
  }

  
// class user {
//     id: number;
//     firstName: string;
//     lastName: string;
//     email: string;
//     gender: string;
//     yearOfBorn: number;
  
//     constructor(data: userData) {
//       this.id = data.id;
//       this.firstName = data.first_name;
//       this.lastName = data.last_name;
//       this.email = data.email;
//       this.gender = data.gender;
//       this.yearOfBorn = data.yearOfBorn;
//     }
//   }

  class UserAccount {
    userId: number;
    operations: Operation[];
    firstName: string;
    lastName: string;
    assignedCards: AssignedCards[]
  
    constructor(data: UserData) {
      this.userId = data.id;
      this.operations = data.operations.map((item) => new Operation(item));
      this.firstName = data.first_name;
      this.lastName = data.last_name;
      this.assignedCards = data.assignedCards.map((item) => new AssignedCards(item));
    }

    getBalance() {
        return this.operations.reduce((prev, curr) => {
            prev += curr.amount;
            return prev
        }, 0)
    }
  }

  class AssignedCards {
      id: number;
      pin: number;
      constructor(data: AssignedCardsData){
          this.id = data.id
          this.pin = data.pin
      }
  }

  class Operation {
    id: number;
    date: Date;
    amount: number;
    constructor(data: OperationData) {
      this.id = data.id;
      this.date = data.date;
      this.amount = data.amount;
    }
  }

  class Card {
    id: number
    CSV_code: number
    firstName: string
    lastName: string
    exp_date: Date

    constructor (id: number, 
        csv: number, 
        name: string,
        surname: string,
        exp_date: Date){
        this.id = id
        this.CSV_code = csv
        this.firstName = name
        this.lastName = surname
        this.exp_date = exp_date
    }
}

class BankProvider {

    private static instance: BankProvider;

    private accountsData!: UserAccount[]
    private data: UserData[];

    public static getInstance(): BankProvider {
        if (!BankProvider.instance) {
            BankProvider.instance = new BankProvider();
        }
        return BankProvider.instance;
    }

    private constructor(){
        this.data = readData();
        this.refreshUsersAccounts();
    }

    public checkPin(
        cardNumber: number,
        pin: number
    ): boolean {
        return !!this.findAccount(cardNumber, pin);
    }

    private findAccount(cardNumber: number,
                        pin: number): UserAccount | undefined {
        return this.accountsData.find(item => {
            return item.assignedCards.find(card => card.id === cardNumber && card.pin === pin)
        });
    }
    private findAccountData(id: number): UserData | undefined {
        return this.data.find(item => item.id === id);
    }

    public payOut(
        cardNumber: number,
        pin: number,
        amount: number
    ){
        const account = this.findAccount(cardNumber, pin);
        if (account) {
            if(account.getBalance() >= amount){

                let operationData: OperationData = {id: account.operations.length+1, date:new Date, amount:-amount}

                account.operations.push(new Operation(operationData))

                const accountData = this.findAccountData(account.userId);
                if (accountData) {
                    accountData.operations.push(operationData);
                    writeData(this.data)
                }

                return true
            } else {
                console.log(`niewystarczajace srodki \n na koncie: ${account.getBalance()} \n do wyplacenia: ${amount}` )
            }
        } else {
            console.log("invalid authorisation")
        }

        return false;
    }

    public checkeBalanceFromBank(
        cardNumber: number,
        pin: number
    ): number | null {
        const account = this.findAccount(cardNumber, pin);
        if (account) {
            return account.getBalance();
        } else {
            console.log("invalid authorisation")
        }
        return null;
    }

    private refreshUsersAccounts() {
        this.accountsData = this.data.map((item:UserData) => new UserAccount(item));
    }
}

class CashMachine {
    private bankProvider: BankProvider
    private stateOfMoney: number
    private inputedCard?: Card
    private inputedPin?: number
    private imputedAmount?: number
    private isLogged = false;
    private maintenanceLogingCode: number

    constructor (stateOfMoney:number, maintenanceCode:number){
        this.stateOfMoney = stateOfMoney
        this.bankProvider = BankProvider.getInstance()
        this.maintenanceLogingCode = maintenanceCode
    }


    public insertCard(card: Card){
        if(this.inputedCard){
            return console.log("Error")
        } else {
            this.inputedCard = card
        }
    }

    public insertPin(pin: number){
        this.inputedPin = pin
    }
    
    public logIn(){
        if (this.checkPinInBank()){
            console.log("you are succesfully logged")
            this.isLogged = true
        }
    }

    public insertAmount(amount: number){
        if (this.isLogged){
                if(this.stateOfMoney >=amount ){
                this.imputedAmount = amount
                console.log(`cash to withrdaw: ${amount}`)
                }
                else console.log(`sorry atm have not enought cash to withdraw`)
            }
            else console.log(`you must first log in`)
    }

    public payOut(){
        if(this.isLogged){
            if (this.imputedAmount && this.payOutFromBank()){
                    this.stateOfMoney -= this.imputedAmount
                    console.log(`You wirdawed ${this.imputedAmount}`)
            }
            
        }
        else console.log(`you must first log in`)
    }

    public maintenance(code: number, maintenance: number){
        if (this.maintenanceLogingCode === code){
            if(maintenance === 1){
                this.checkAtmStatus()
            }
            else if(maintenance === 2){
                console.log(`opening cashbox`)
            }
            else if(maintenance === 3){
                console.log(`restaring...`)
            }
            else console.log(`comend unrecognized`)

        }
        else console.log(`wrong maintenence code`)
    }

    public checkBalance(){
        if(this.isLogged){
            if(this.inputedCard?.id && this.inputedPin){
                const saldo = this.bankProvider.checkeBalanceFromBank(this.inputedCard.id, this.inputedPin)
                if (saldo === null) {
                    console.log(`Unknown saldo in your bank account`)
                } else {
                    console.log(`You have ${saldo} money in your bank account`)
                }
            }
        } else {
            console.log(`you must first insert amout of money you want withdraw`)
        }
    }
    
    private checkAtmStatus(){
        return console.log(`money left in casbox ${this.stateOfMoney}`)
    }

    private checkPinInBank(){
        if(this.inputedCard?.id && this.inputedPin){
            return this.bankProvider.checkPin(this.inputedCard.id, this.inputedPin)
        } else {
            console.log(`you must first insert amout of money you want withdraw`)
        }
    }

    private payOutFromBank(){
        if(this.inputedCard?.id && this.inputedPin && this.imputedAmount){
        return this.bankProvider.payOut(this.inputedCard.id, this.inputedPin, this.imputedAmount)
        }
    }
}

const atm = new CashMachine(10000, 12345)

const card1 = new Card(1001, 12, "Jan", "Kowalski", new Date ())

atm.insertCard(card1)

atm.insertPin(1234)

atm.logIn()

atm.checkBalance()

atm.insertAmount(199)

atm.payOut()
//
// atm.checkBalance()

// atm.maintenance(12345, 1)

// atm.checkMoneyInATM()

// const userAccount1 = new userAccount(data[0]);
// console.log(userAccount1);

// const bankprovider1 = new BankProvider(data)
// if (bankprovider1.payOut(1001, 1234, 10) === true){
//  console.log("AAA")
// }

// console.log(bankprovider1)

// data.map((item:any) => item.assignedCards.forEach((el:assignedCards) => 
//         console.log(el.id) ))
// const a1 = new userAccount(data[0])
// console.log(a1)

File: bankomat/file-version/bankomat.py
import json
import functools
import datetime

def readData():
    with open('accounts.json') as file:
        clients = json.load(file)
    return clients


def writeData(listObj):
    with open('accounts.json', 'w') as json_file:
        json.dump(listObj, json_file,
                  indent=4,
                  separators=(',', ': '))


class AssignedCards():
    def __init__(self, data):
        self.number = data["id"]
        self.pin = data["pin"]


class Operation():
    def __init__(self, data):
        self.operationId = data["id"]
        self.operationDate = data["date"]
        self.moneyNum = data["amount"]


class UserAccount():

    def __init__(self, data):

        self.id = data["id"]
        self.name = data["first_name"]
        self.surname = data["last_name"]
        self.assignedCards = list(map(
            lambda item: AssignedCards(item), data["assignedCards"]))
        self.operations = list(map(lambda item: Operation(item), data["operations"]))

    def getBalance(self):
        return functools.reduce(lambda prev, curr: prev + curr.moneyNum, self.operations, 0)


class Card():
    def __init__(self, number, cvs, name, surname, expiryDate):
        self.number = number
        self.cvs = cvs
        self.name = name
        self.surname = surname
        self.expiryDate = expiryDate


class BankProvider(dict):
    def __init__(self):
        self.__data = readData()
        self.accounts = list(map(lambda item: UserAccount(item), self.__data))

    __instance__ = None

    def __new__(cls, *args, **kwargs):
        if BankProvider.__instance__ is None:
            BankProvider.__instance__ = dict.__new__(cls)
        return BankProvider.__instance__

    def __findAccount(self, number, pin):
        founded = None
        #print('start finding', len(self.accounts))
        for account in self.accounts:
            #print('Account: ', account.id)
            for card in account.assignedCards:
                #print(card.number, card.pin, 'vs', number, pin)
                if card.number == number and card.pin == pin:
                    founded = account
                    return account
        #print('not found')
        return founded

    def __findAccountData(self, id):
        founded = list(filter(lambda item: item["id"] == id, self.__data))
        if len(founded) > 0:
            return founded[0]
        else:
            return None

    def checkPin(self, number, pin):
        return self.__findAccount(number, pin) is not None

    def payOut(self, number, pin, money):
        account = self.__findAccount(number, pin)
        if account is not None:
            if account.getBalance() >= money:
                preparedMoney = account.getBalance()
                preparedId = len(account.operations) + 1
                preparedDate = datetime.datetime.now().strftime("%m/%d/%Y, %H:%M:%S")
                preparedMoney = -money
                operationData = {"id": preparedId,
                                 "date": preparedDate, "amount": preparedMoney}
                account.operations.append(Operation(operationData))

                accountData = self.__findAccountData(account.id)
                #print('accountData', accountData)
                if accountData is not None:
                    accountData["operations"].append(operationData)
                    writeData(self.__data)
                    #print('writeData')
                    return True
            else:
                print("Error: not enough money")
        else:
            print("Error: invalid authorisation")
        return False

    def checkBalanceFromBank(self, number, pin):
        account = self.__findAccount(number, pin)
        if account is not None:
            return account.getBalance()
        else:
            print("Error: invalid authorisation")
        return None


class CashMashine():
    def __init__(self, moneyAmount):

        self.__moneyAmount = moneyAmount
        self.__bankProvider = BankProvider()
        self.__inputedCard = None
        self.__inputedPin = None
        self.__inputedAmount = None
        self.__isLogged = False

    def insertCard(self, card):
        if self.__inputedCard != None:
            return "Error"

        self.__inputedCard = card

    def insertPin(self, pin):
        self.__inputedPin = pin

    def __checkPinInBank(self):
        if self.__inputedCard and self.__inputedCard.number and self.__inputedPin:
            return self.__bankProvider.checkPin(self.__inputedCard.number, self.__inputedPin)
        else:
            print("Error: you must first insert amout of money you want withdraw")

    def __payOutFromBank(self):
        if self.__inputedCard and self.__inputedCard.number and self.__inputedPin:
            return self.__bankProvider.payOut(self.__inputedCard.number, self.__inputedPin, self.__inputedAmount)

    def logIn(self):
        if self.__checkPinInBank():
            print("you are succesfully logged")
            self.isLogged = True
        else:
            print("Error: wrong pin or card number")
            self.isLogged = False

        return self.isLogged

    def insertAmount(self, amount):
        if self.isLogged:
            if amount <= self.__moneyAmount:
                self.__inputedAmount = amount
                print(f"cash to withdraw: {amount}")
            else:
                print("Error: sorry atm have not enought cash to withdraw")
        else:
            print("Error: you aren't logged in")

    def payOut(self):
        if self.isLogged:
            if self.__inputedAmount:
                if(self.__payOutFromBank()):
                    self.__moneyAmount -= self.__inputedAmount
                    print(f"you withdrew {self.__inputedAmount}")
                else:
                    print("Pay out from bank error")
            else:
                print("Please choose amount")

        else:
            print("you aren't logged in")

    def checkBalance(self):
        if self.isLogged:
            if self.__inputedCard.number and self.__inputedPin:
                balance = self.__bankProvider.checkBalanceFromBank(
                    self.__inputedCard.number, self.__inputedPin)
                if balance == None:
                    print("Unknown saldo in your bank account")
                else:
                    print(f"You have {balance} money in your bank account")
            else:
                print("Error: firstly you must insert amount of money you want to withdraw")

atm = CashMashine(10000000000000)
card1 = Card(1001, 12, "Jan", "Kowalski", datetime.datetime.now())
atm.insertCard(card1)
atm.insertPin(1234)
if atm.logIn():
    atm.checkBalance()
    atm.insertAmount(100)
    atm.payOut()
    print("Thank you")
